stage1/init/common: GenerateMounts refactoring

ybubnov opened the pull request:
This patch provides a refactoring of the GenerateMounts function from
the "stage1/init/common" package in order to reduce the amount of the
loops over the mount points.

The implementation replaces the old function "GenerateMounts" and ships
a new type called "RuntimeMounts" that provides "Mounts" to reproduce
previous behavior and a new function "MountsFunc" that calls a
user-defined function for each mount point.

closes #2380

ybubnov committed:
stage1/init/common: GenerateMounts refactoring

This patch provides a refactoring of the GenerateMounts function from
the "stage1/init/common" package in order to reduce the amount of the
loops over the mount points.

The implementation replaces the old function "GenerateMounts" and ships
a new type called "RuntimeMounts" that provides "Mounts" to reproduce
previous behavior and a new function "MountsFunc" that calls a
user-defined function for each mount point.

lucab commented:
ok to test

lucab commented:
note to self: this embeds a copy of https://github.com/coreos/rkt/pull/3590, see tacking in there.

ybubnov commented:
@alban, @lucab, apologies for being too noisy. I just wanted to ask if these changes are useful, as I didn't get any feedback.

lucab commented:
Sorry for the delay, we have been traveling a bit recently and still going through backlog. I'll try to have a review pass on this soon, in the meanwhile I think the test change (in rkt_non_root_test.go) is superfluous now.

ybubnov commented:
@lucab, thank you, I will rebase a commit with the latest changes in the master branch.

ybubnov commented:
Not sure if the failures in the ```debian-testing``` caused by introduces changes, any help appreciated.

squeed commented on stage1/init/common/mount.go:
```diff
@@ -68,131 +69,241 @@ func (m *Mount) Source(podRoot string) string {
 
-	// Map of hostpath -> Mount
-	mnts := make(map[string]schema.Mount)
+// mountsMap is a mapping of mount point path to mount point details.
```
Minor doc issue: this is a mapping from mount point name, right?

ybubnov replied:
The ```name``` in this context means a file-system path, like ```/mnt/dir1```, this is why I decided to make it clear and call it a "path". If the "mount point name" is preferable I will change it.

squeed replied:
Ah, I see. the fact that it's a `types.ACName` is a bit confusing, as UNIX paths can have more characters than an ACName.

ybubnov replied:
@squeed, sorry, it was my incorrect understanding, I looked closely at the manifest specification, and this is really a name of the mount point, not a path:

```json
"mountPoints": [
  {
    "name": "work",
    "path": "/var/lib/work",
    "readOnly": false
  }
]
```

Source: https://github.com/appc/spec/blob/master/spec/aci.md I will update the comments accordingly.

lucab commented:
I think this PR is mostly fine from my point of view, but I'll wait for @squeed to come back for the final review.

squeed commented on stage1/init/common/mount.go:
```diff
@@ -68,131 +69,241 @@ func (m *Mount) Source(podRoot string) string {
+
+	// processed defines a set of processed mount points, so we don't
+	// have to walk though them multiple times.
+	processed map[string]struct{}
```
This seems to indicate that RuntimeMounts is essentially only iterable once.

That makes me a bit uncomfortable - this sort of subtle side-effecting behaviour - I could easily see another developer calling MountsFunc() more than once.

Either RuntimeMounts should error / panic if called twice, or should be non-side-effecting.

ybubnov replied:
The idea was to decouple the implementation into multiple small functions, but it resulted in a need to share the state between them. Thereby I moved this state into a structure, but now it really makes more sense to define a separate type to bypass that state, in order to get rid of undesirable side-effect.

squeed commented:
Can you add some documentation to the `mounted` and `nonmounted` functions - their usage isn't quite clear.

It does make me uncomfortable to have so hidden state - the "processed" array is a bit "spooky". Would you mind documenting its use as a shortcut?

ybubnov commented:
Sure, I will update the documentation with more details.

ybubnov commented:
I've moved the state to a separate type, which is created for each call of both ```Mounts``` and ```MountsFunc```, such that it does not lead to a side-effect.

Additionally I've renamed functions ```mounted``` and ```nonmounted``` to ```traverseMounts``` and ```traverseMountPoints``` respectively (I hope, such names better describe their purposes). Have updated the documentation with more details.

ybubnov commented:
For some reason ```semaphoreci``` does not pick up a job, I will try to re-open a pull-request to see if it helps.

lucab commented:
Rebasing on top of current master (ie. https://github.com/rkt/rkt/pull/3670) would probably help on the testing side.

ybubnov commented:
@lucab, thank you! I'll give it a try.

euank commented on stage1/init/common/mount.go:
```diff
@@ -68,131 +69,263 @@ func (m *Mount) Source(podRoot string) string {
-				UID:  &defaultUID,
-				GID:  &defaultGID,
-			}
+			log.Printf("warning: no volume specified for mount point "+
```
I prefer not breaking long loglines to make sure they remain greppable.

ybubnov replied:
I've squashed comment into a single line.

euank reviewed (commented):
Other than my logline nit, this looked good to me.

lucab commented:
It seems fine overall. @squeed can you please have a final pass on this?

squeed commented on stage1/init/common/mount.go:
```diff
@@ -68,131 +69,260 @@ func (m *Mount) Source(podRoot string) string {
-				break
-			}
-		}
+		mpoint, _ := s.mountPointOf(m.Volume)
```
What happens if mpoint is null?

ybubnov replied:
This function returns instance of ```types.MountPoint```, which is a structure and default value for it is an *empty* structure. Therefore, when mount point was not found, it returns ```types.MountPoint{}```, where ```ReadOnly``` attribute is set to ```false``` - a default value for boolean type.

The answer is ```mpoint``` is never going to be equal to ```nil```.

ybubnov commented:
Hi, do I need to perform any additional changes or to explain unclear implementation? The review got stuck for a week, apologies for being too noisy.
